### PR TITLE
[CI] Fix two flaky Windows tests (root causes, not skips)

### PR DESCRIPTION
**Pull request by Wauplin**
### Context: let's hope CI is finally green :crossed_fingers: 

---

Two Windows-only flaky tests, both with real root causes:

### 1. `test_safetensors_index_rejects_absolute_path` (Py 3.14 Windows)

This is a real security regression, not just a test bug. The validator in `_load_sharded_checkpoint` uses `os.path.isabs(shard_file)` to reject absolute paths in a safetensors index file. Since **Python 3.13**, `ntpath.isabs()` no longer treats `/tmp/foo` (single leading slash, no drive) as absolute — so on Windows Py 3.13+, an index file with `"/tmp/malicious.safetensors"` sails past validation and `os.path.join` resolves it to `C:\tmp\malicious.safetensors`.

Fix: check both `PurePosixPath.is_absolute()` and `PureWindowsPath.is_absolute()`, so a malicious POSIX-absolute path in an index file is rejected on Windows too (and vice-versa). Parametrized the test to cover POSIX-absolute, Windows drive-absolute, and UNC paths.

### 2. `test_hf_file_system.test_pickle` (Windows, all Py versions)

The test did `HfFileSystem()` (defaults to **production**) but `self.text_file` is a repo created on **staging**. The `isfile()` call always 404'd on production. In `resolve_path`, when a lookup fails with `RepositoryNotFoundError`, the code retries with a truncated repo_id (namespace-fallback) — writing a second cache entry. Whether the retry fires depends on the exact error returned by production, which differs between GHA Ubuntu and Windows runners. Windows consistently gets a clean 404 → fallback fires → 2 entries → `assert len(...) == 1` fails. Linux gets a different error → 1 entry → passes.

Fix: point the pickled fs at `ENDPOINT_STAGING` so the initial `isfile()` resolves deterministically against the endpoint where the repo actually exists.

## Verification

```
$ pytest tests/test_serialization.py::TestShardedCheckpointValidation
...
tests/test_serialization.py::TestShardedCheckpointValidation::test_safetensors_index_rejects_bin_shard PASSED
tests/test_serialization.py::TestShardedCheckpointValidation::test_safetensors_index_rejects_path_traversal PASSED
tests/test_serialization.py::TestShardedCheckpointValidation::test_safetensors_index_rejects_absolute_path[/tmp/malicious.safetensors] PASSED
tests/test_serialization.py::TestShardedCheckpointValidation::test_safetensors_index_rejects_absolute_path[C:\\malicious.safetensors] PASSED
tests/test_serialization.py::TestShardedCheckpointValidation::test_safetensors_index_rejects_absolute_path[\\\\server\\share\\malicious.safetensors] PASSED
5 passed
```

Windows CI behavior will be validated by this PR's own checks.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches shard filename validation in model loading (a security-sensitive area) and adjusts filesystem caching tests; logic change is small but could affect loading behavior for unusual paths across OSes.
> 
> **Overview**
> **Hardens sharded safetensors checkpoint loading** by validating shard filenames using both `PurePosixPath.is_absolute()` and `PureWindowsPath.is_absolute()` (in addition to rejecting `..`), closing a Windows/Python 3.13+ edge case where POSIX-style absolute paths could bypass `os.path.isabs` checks.
> 
> **Deflakes Windows CI tests** by (1) parametrizing the absolute-path rejection test to cover POSIX, Windows drive, and UNC paths on any host OS, and (2) making `HfFileSystem` pickling/caching test use the staging endpoint so cache population is deterministic.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8c7756ba33d4f24de02f303ae807dd06f2d496b2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/huggingface_hub/serialization/_torch.py [hunk 1]
```diff
@@ -516,8 +516,15 @@ def _load_sharded_checkpoint(
...
-        if os.path.isabs(shard_file) or ".." in Path(shard_file).parts:
+        # Reject path traversal (e.g. "../malicious.bin", absolute paths).
+        # Check both POSIX and Windows semantics so e.g. "/tmp/x" is rejected on
+        # Windows too. Note: `os.path.isabs` is host-OS-specific, and since Python 3.13
+        # ntpath.isabs no longer treats a single leading "/" as absolute.
+        if (
+            PurePosixPath(shard_file).is_absolute()
+            or PureWindowsPath(shard_file).is_absolute()
+            or ".." in Path(shard_file).parts
+        ):
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/serialization/_torch.py]: Good catch — that's a real gap. Replaced the `PurePosixPath.is_absolute() or PureWindowsPath.is_absolute()` combo with `win_path.drive or win_path.root or '..' in win_path.parts`, which rejects Windows rooted-without-drive (`\\evil`), drive-only (`C:evil`), UNC, POSIX-absolute, and both `../` and `..\\` traversal in a single expression. Added parametrized tests for the new cases.

**Comment by Wauplin** [comment]
@hanouticelina ready for review. To be honest, not a PR from me but the comments look sensible to me. And since CI is passing I think it's good to merge.

**Review by hanouticelina — APPROVED** [review]
The PR description and the comments make sense to me! 
thanks for taking care of that! (I don't remember the last time Windows tests passed 😄)
